runtime-modules/referendum/src/lib.rs

```diff
@@ -0,0 +1,740 @@
+// TODO: module documentation
+// TODO: adjust all extrinsic weights
+
+// NOTE: This module is instantiable pallet as described here https://substrate.dev/recipes/3-entrees/instantiable.html
+// No default instance is provided.
+
+/////////////////// Configuration //////////////////////////////////////////////
+#![allow(clippy::type_complexity)]
+#![cfg_attr(not(feature = "std"), no_std)]
+
+// used dependencies
+use codec::{Codec, Decode, Encode};
+use frame_support::{
+    decl_error, decl_event, decl_module, decl_storage, error::BadOrigin, traits::Get, Parameter,
+    StorageValue,
+};
+use sp_arithmetic::traits::{BaseArithmetic, One};
+use sp_runtime::traits::{MaybeSerialize, Member};
+use std::marker::PhantomData;
+use system::ensure_signed;
+
+use std::collections::HashSet;
+
+// declared modules
+mod mock;
+mod tests;
+
+/////////////////// Data Structures ////////////////////////////////////////////
+
+#[derive(Encode, Decode, PartialEq, Eq, Debug)]
+pub enum ReferendumStage {
+    Void,
+    Voting,
+    Revealing,
+}
+
+impl Default for ReferendumStage {
+    fn default() -> ReferendumStage {
+        ReferendumStage::Void
+    }
+}
+
+#[derive(Encode, Decode, PartialEq, Eq, Debug, Default)]
+pub struct SealedVote<Hash, CurrencyBalance> {
+    commitment: Hash,
+    stake: CurrencyBalance,
+}
+
+#[derive(Clone, Encode, Decode, PartialEq, Eq, Debug)]
+pub enum ReferendumResult<ReferendumOption, VotePower> {
```

Comment:
I don't know whether you actually need this type, my next deeper review will look into this, however, if you want to represent outcomes, this is not the best type. Ill leave it for you to review based on your final rewrite.

Reply:
This enum handles all possible situations referendum can result in:
- there are X winners as requested
- X winners were expected but Xth winner has the same amount of votes as (X+1)th option
- X winners were expected but only Y (`Y < X`) options received any votes
- nobody reveals a valid vote in referendum

Reply:
The issue continues here - https://github.com/Joystream/joystream/pull/1133#discussion_r477369394